src/Features/LanguageServer/Protocol/CustomProtocol/FindUsagesLSPContext.cs

```diff
@@ -0,0 +1,220 @@
+            new Dictionary<DefinitionItem, int>();
+
+        private readonly List<VSReferenceItem> _resultsChunk =
+            new List<VSReferenceItem>();
+
+        private readonly IProgress<object[]> _progress;
+        private readonly Document _document;
+        private readonly int _position;
+        private readonly IMetadataAsSourceFileService _metadataAsSourceFileService;
+
+        public override CancellationToken CancellationToken { get; }
+
+        public FindUsagesLSPContext(
+            IProgress<object[]> progress,
+            Document document,
+            int position,
+            IMetadataAsSourceFileService metadataAsSourceFileService,
+            CancellationToken cancellationToken)
+        {
+            _progress = progress;
+            _document = document;
+            _position = position;
+            _metadataAsSourceFileService = metadataAsSourceFileService;
+
+            CancellationToken = cancellationToken;
+        }
+
+        public async override Task OnCompletedAsync()
+        {
+            var results = Array.Empty<VSReferenceItem>();
+            using (await _semaphore.DisposableWaitAsync(CancellationToken).ConfigureAwait(false))
+            {
+                if (_resultsChunk.IsEmpty())
+                {
+                    return;
+                }
+
+                results = _resultsChunk.ToArray();
+                _resultsChunk.Clear();
+            }
+
+            _progress.Report(results);
+        }
+
+        public async override Task OnDefinitionFoundAsync(DefinitionItem definition)
+        {
+            using (await _semaphore.DisposableWaitAsync(CancellationToken).ConfigureAwait(false))
+            {
+                if (!_definitionToId.ContainsKey(definition))
+                {
+                    // Assinging a new id to the definition
+                    _definitionToId.Add(definition, _id);
+
+                    // VSReferenceItem currently doesn't support the ClassifiedTextElement type for DefinitionText,
+                    // so for now we just pass in a string.
+                    // https://devdiv.visualstudio.com/DevDiv/_workitems/edit/1100009
+                    var classifiedText = definition.GetClassifiedText();
+                    var definitionText = "";
+                    foreach (var text in classifiedText.Runs)
+                    {
+                        definitionText += text.Text;
+                    }
+
+                    // Creating a new VSReferenceItem for the definition
+                    var definitionItem = await GenerateVSReferenceItemAsync(
+                        _id, definitionId: _id, _document, _position, definition.SourceSpans.FirstOrDefault(),
+                        definition.DisplayableProperties, _metadataAsSourceFileService, definitionText,
+                        symbolUsageInfo: null, CancellationToken).ConfigureAwait(false);
+
+                    if (definitionItem.Location != null)
```

Comment:
curious when .Location would be null.  if we're not going to use the item without a location, then should GenerateVSReferenceItemAsync just return null in that case?